packages/test/test-end-to-end-tests/src/test/testApi.ts

```diff
@@ -0,0 +1,240 @@
+import { ConsensusQueue } from "@fluidframework/ordered-collection";
+import { ConsensusRegisterCollection } from "@fluidframework/register-collection";
+import { SharedString, SparseMatrix } from "@fluidframework/sequence";
+import { TestFluidObjectFactory } from "@fluidframework/test-utils";
+
+// ContainerRuntime and Data Runtime API
+import { ContainerRuntimeFactoryWithDefaultDataStore, DataObject, DataObjectFactory } from "@fluidframework/aqueduct";
+
+import { lock } from "proper-lockfile";
+import * as semver from "semver";
+import { pkgVersion } from "./packageVersion";
+
+// List of package that needs to be install for past versions
+const packageList = [
+    "@fluidframework/aqueduct",
+    "@fluidframework/test-utils",
+    "@fluidframework/container-loader",
+    "@fluidframework/container-runtime",
+    "@fluidframework/cell",
+    "@fluidframework/counter",
+    "@fluidframework/ink",
+    "@fluidframework/map",
+    "@fluidframework/matrix",
+    "@fluidframework/ordered-collection",
+    "@fluidframework/register-collection",
+    "@fluidframework/sequence",
+];
+
+// Current versions of the APIs
+const LoaderApi = {
+    version: pkgVersion,
+    Loader,
+};
+
+const ContainerRuntimeApi = {
+    version: pkgVersion,
+    ContainerRuntime,
+    ContainerRuntimeFactoryWithDefaultDataStore,
+};
+
+const DataRuntimeApi = {
+    version: pkgVersion,
+    DataObject,
+    DataObjectFactory,
+    TestFluidObjectFactory,
+    dds: {
+        SharedCell,
+        SharedCounter,
+        Ink,
+        SharedDirectory,
+        SharedMap,
+        SharedMatrix,
+        ConsensusQueue,
+        ConsensusRegisterCollection,
+        SharedString,
+        SparseMatrix,
+    },
+};
+
+export type DataRuntimeApiType = typeof DataRuntimeApi;
+const resolutionCache = new Map<string, string>();
+
+function resolveVersion(requested: string) {
+    const cachedVersion = resolutionCache.get(requested);
+    if (cachedVersion) { return cachedVersion; }
+    if (semver.valid(requested)) {
+        resolutionCache.set(requested, requested);
+        return requested;
+    }
+
+    const result = execSync(
+        `npm v @fluidframework/container-loader@"${requested}" version --json`,
+        { encoding: "utf8" },
+    );
+    try {
+        const versions: string | string[] = JSON.parse(result);
+        const version = Array.isArray(versions) ? versions.sort(semver.rcompare)[0] : versions;
+        if (!version) { throw new Error(`No version found for ${requested}`); }
+        resolutionCache.set(requested, version);
+        return version;
+    } catch (e) {
+        throw new Error(`Error parsing versions for ${requested}`);
+    }
+}
+
+async function ensureInstalled(requested: string) {
+    const version = resolveVersion(requested);
+    let release = await lock(__dirname, { retries: { forever: true } });
+    try {
+        // Assuming this file is in dist\test, so go to ..\..\node_modules\.legacy as the install location
```

Comment:
can't we just use checkInstalled here

Reply:
checkInstalled throws.  We don't want to throw here.  But I will refactor the path creation into a function.

Reply:
Done